jetty-server/src/main/java/org/eclipse/jetty/server/ContentProcessor.java

```diff
@@ -0,0 +1,205 @@
+{
+    private final AutoLock _lock = new AutoLock();
+    private boolean _reading;
+    private boolean _iterating;
+    private boolean _available;
+    private Content _output;
+    private Runnable _onContentAvailable;
+
+    public ContentProcessor(Content.Provider provider)
+    {
+        super(provider);
+    }
+
+    @Override
+    public Content readContent()
+    {
+        boolean available;
+        try (AutoLock ignored = _lock.lock())
+        {
+            if (_reading)
+                return null;
+
+            if (_output != null)
+            {
+                Content output = _output;
+                _output = null;
+                return output;
+            }
+
+            available = _available;
+            _available = false;
+            _reading = true;
+        }
+
+        try
+        {
+            while (true)
+            {
+                Content output = available ? null : produce(null);
+                if (output != null)
+                    return output;
+                available = false;
+                Content input = getProvider().readContent();
+                output = produce(input);
+                if (output != null)
+                    return output;
+                if (input == null)
+                    return null;
+            }
+        }
+        finally
+        {
+            try (AutoLock ignored = _lock.lock())
+            {
+                _reading = false;
+            }
+        }
+    }
+
+    @Override
+    public void demandContent(Runnable onContentAvailable)
+    {
+        Objects.requireNonNull(onContentAvailable);
+        iterate(false, onContentAvailable);
+    }
+
+    private void iterate(boolean available, Runnable onContentAvailable)
+    {
+        try (AutoLock ignored = _lock.lock())
+        {
+            if (_reading)
+                throw new IllegalStateException("read pending");
+
+            if (onContentAvailable != null)
+            {
+                if (_onContentAvailable != null && onContentAvailable != _onContentAvailable)
+                    throw new IllegalStateException("demand pending");
+                _onContentAvailable = onContentAvailable;
+                onContentAvailable = null;
+            }
+
+            _available |= available;
+
+            if (_iterating)
+                return;
+
+            if (_output == null)
+            {
+                available = _available;
+                _available = false;
+            }
+            else
+            {
+                onContentAvailable = _onContentAvailable;
+                _onContentAvailable = null;
+            }
+
+            _iterating = true;
+        }
+
+        Content output = null;
+        boolean iterating = true;
+        while (iterating)
+        {
+            if (onContentAvailable != null)
+            {
+                try
+                {
+                    onContentAvailable.run();
```

Comment:
I'm not sure we can ever really support any `cancel` semantic as you can't take back demand.

But your point is valid - we need to better test error handling in the `Processor` and probably in `Provider` generally.    I've got some tests of error cases for the direct usage of `Provider` in the `Request` API, but probably not enough.... and none in this PR.

I'll add the simple cases of an exception thrown from an 'onContentAvailable` callback, and also one thrown from `produce`.   If you can think of any other specific error use-cases you want tested... and the outcome you expect, then do either write them or describe them here.

In my other handling, I have tried to avoid additional error channel and instead have kept errors "in-band", so inboud errors are reported via a `readContent` and outbound errors are signalled by a `Callback.failed(x)` on the original request invocation.    Any `onError` methods just invite all the complexities of concurrent invocation and errors thrown from `onError` etc.          But let's do the testing here and see how it looks.

Reply:
@sbordet @lorban I've pushed an update that now handles errors thrown from `process` or `available` calls.   It handles all errors inband, by providing them via `readContent()`.  If the consumer is not reading nor demanding, then the error is held until they are.     

To make this sensible, I've introduced the concept of a non-last error!  So you can read an `Content.Error` from `readContent()`, that if it did not originate from the original provider will not be last (unless the input that caused the error to be generated was already last).

There is some risk of applications infinitely looping if in their available method they: 1) readContent; 2) demandContent; 3) throw.   But this is really no different to the looping danger if an onError callback throws.  I could add loop protection?   Actaully I have and idea for that..... stand by....

Reply:
hmmm no that idea didn't work. anyway have a look and feedback please.